### PR TITLE
Fixed dns lookup in case of server close connection with econnrefused

### PR DESCRIPTION
**Pull request by egobrain**
In case of ns server responded with {error, econnrefused} we should close udp connection and start new one for next ns server.

In my case there is `/etc/resolve.conf` with folowing structure

```
nameserver 127.0.0.1
nameserver 8.8.8.8
search ghcg.com
```

and i tried to resolve an srv record, but local dns server (127.0.0.1) crashed on request.

```
$ dig srv _tickdb-qida-ws._tcp.load.ghcg.com

; <<>> DiG 9.10.3-P4-Ubuntu <<>> srv _tickdb-qida-ws._tcp.load.ghcg.com
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 55939
;; flags: qr rd ra; QUERY: 1, ANSWER: 1, AUTHORITY: 4, ADDITIONAL: 7

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 1280
;; QUESTION SECTION:
;_tickdb-qida-ws._tcp.load.ghcg.com. IN SRV

;; ANSWER SECTION:
_tickdb-qida-ws._tcp.load.ghcg.com. 300 IN SRV  1 0 8585 tickdb3.load.zorg.sh.

;; AUTHORITY SECTION:
load.ghcg.com.          300     IN      NS      ns-980.awsdns-58.net.
load.ghcg.com.          300     IN      NS      ns-1638.awsdns-12.co.uk.
load.ghcg.com.          300     IN      NS      ns-235.awsdns-29.com.
load.ghcg.com.          300     IN      NS      ns-1395.awsdns-46.org.

;; ADDITIONAL SECTION:
ns-235.awsdns-29.com.   84551   IN      A       205.251.192.235
ns-235.awsdns-29.com.   81987   IN      AAAA    2600:9000:5300:eb00::1
ns-980.awsdns-58.net.   28641   IN      A       205.251.195.212
ns-980.awsdns-58.net.   21916   IN      AAAA    2600:9000:5303:d400::1
ns-1395.awsdns-46.org.  16772   IN      A       205.251.197.115
ns-1395.awsdns-46.org.  90528   IN      AAAA    2600:9000:5305:7300::1

;; Query time: 188 msec
;; SERVER: 127.0.1.1#53(127.0.1.1)
;; WHEN: Fri Sep 01 11:35:46 MSK 2017
;; MSG SIZE  rcvd: 372
```

```
1> inet_res:getbyname("_tickdb-qida-ws._tcp.load.ghcg.com", srv).                                                 
{error, timeout}.
```

In case of first ns server crashes erlang should use next dns server (8.8.8.8), but request crashes on erlang side with {error,einval} badmatch.

Here is simple example what's going on if we dont close connection after first {error, econnrefused}:

```erlang 
1> Test = fun(Urls) ->
    Buffer = <<178,7,1,0,0,1,0,0,0,0,0,0,17,95,116,105,99,107,100,98,45,113,119,101,98,45,
        104,116,116,112,4,95,116,99,112,4,108,111,97,100,4,103,104,99,103,3,99,111,
        109,0,0,33,0,1>>,
    T = 5000,
    {ok, S} = gen_udp:open(0, [{active,false},binary,inet]),
    lists:map(fun({IP, Port}=Url) ->
        ok = gen_udp:connect(S, IP, Port),
        ok = gen_udp:send(S, IP, Port, Buffer),
        {Url, gen_udp:recv(S, 0, T)}
    end, Urls)
end.

2> Test([{{127,0,0,1}, 53}, {{8,8,8,8}, 53}]).
** exception error: no match of right hand side value {error,einval}

3> Test([{{8,8,8,8}, 53}]).                   
[{{{8,8,8,8},53},
  {ok,{{8,8,8,8},
       53,
       <<178,7,129,128,0,1,0,1,0,0,0,0,17,95,116,105,99,107,
         100,98,45,...>>}}}]

4> Test([{{8,8,8,8}, 53}, {{127,0,0,1}, 53}]).
[{{{8,8,8,8},53},
  {ok,{{8,8,8,8},
       53,
       <<178,7,129,128,0,1,0,1,0,0,0,0,17,95,116,105,99,107,
         100,98,45,...>>}}},
 {{{127,0,0,1},53},{error,econnrefused}}]
```

The solution is to close udp connection on {error, econnrefused} message

### CODE REVIEW

File: lib/kernel/src/inet_res.erl [hunk 1]
```diff
@@ -859,15 +859,17 @@ query_ns(S0, Id, Buffer, IP, Port, Timer, Retry, I,
...
 			    S, Id, Buffer, IP, Port, Timeout, Verbose) of
 			 {ok,#dns_rec{header=H}} when H#dns_header.tc ->
 			     TcpTimeout = inet:timeout(Tm*5, Timer),
-			     query_tcp(
-			       TcpTimeout, Id, Buffer, IP, Port, Verbose);
-			 Reply -> Reply
-		     end};
+			     {S, query_tcp(
+			       TcpTimeout, Id, Buffer, IP, Port, Verbose)};
+			{error, Reason}=Err when Reason =:= econnrefused ->
```
Comment (RaimoNiskanen) [inline lib/kernel/src/inet_res.erl]: It is neater with `{error, econnrefused} = Err ->`

**Comment by RaimoNiskanen** [comment]
Looks right to me

File: lib/kernel/src/inet_res.erl (hunk 1 above)
Comment (egobrain) [reply lib/kernel/src/inet_res.erl]: Fixed

**Comment by RaimoNiskanen** [comment]
Merged.  Thank you for your contribution!

**Pull request merged**